### ISSUE TITLE
[SC] Allow user to set scaling threshold (for COS)

### ISSUE THREAD

**Issue by oyvindeide** [equinor/semeio#47]
When running the correlated observations scaling after the spearman correlation, the threshold is hard-coded to 0.95. Allow the user to modify this value, e.g. through an argument

Including many arguments may justify creating an optional config file:
https://github.com/equinor/semeio/issues/46
https://github.com/equinor/semeio/issues/44

**Comment by jondequinor** [comment]
I think adding an additional configuration file to the SC/COS domain would serve only to confuse the user. It certainly confuses me. SC already calls COS, now configuring COS through SC is bad design. A higher level of abstraction is needed, one configuration to serve (at least) the following cases:

- SC only
- COS only
- SC then COS

So I think we need a new configuration schema that allows the user to express these. Any thoughts?

**Comment by oyvindeide** [comment]
I´ll try to expand a bit on rational behind the current setup. The original scaling job was not frequently used outside of research, and they wanted a workflow that was as automated as possible with a minimal setup with good defaults for the average user. That is why they´re running back-to-back. Now, a setup like what you´re describing does make sense, we just have to make sure we´re not just making another `ert` config (as the approach you´re sketching mirrors much of the `ert` config we´re trying to make in the future). This is kind of already possible in the current `ert` config, it´s just a bit difficult to set up as you have to have good knowledge of how the jobs are operating. A feature of whatever we come up with is that it has to be flexible enough to allow the output of the `spearman` job be the input of the `scaling` job.

**Comment by sondreso** [comment]
The configuration discussion here is very related to the discussion we have had earlier about allowing configuration through the plugin system. However, I guess you need a solution to this before that system will be implemented.

**Comment by jondequinor** [comment]
@sondreso yes! Do you remember if we had anything about that in writing? We did decide to move forward with that work, so maybe it can be started here.

@oyvindeide SC can internally call COS directly, this configuration would not constrain implementation.

Re: recreating the new ERT config—I am not too concerned about this. If we capture intent over process, it is trivial for both us and the user to migrate to the prophesized ert config.

**Comment by sondreso** [comment]
Nothing in writing I think. We decided that that was the direction we wanted to go, but the work was left  in favor of fixing komodo. There was talk of this related to the STEA pr that @xjules had.

**Comment by jondequinor** [comment]
@markusdregi also wanted to comment on this issue.

**Comment by markusdregi** [comment]
**Note: This is only a draft. I'll finish it tomorrow morning.**

I think this would be a great first demonstration of a stage of a model updating pipeline that could later be lifted into the general ERT configuration file. In the end, what I would like to see is a data pipeline to describe the updating of a model in a classic pipes and filter design.

I'll refer to the main pipeline as the model updating pipeline, which consists of multiple stages. Each stage again can define input and output records (data streams), records local to the scope of the stage and a pipeline. Each stage pipeline consists of a list of functions, where each function can have records local to their scope.

The records input to the model updating pipeline at large would be the `parameters`, `simulation_results` and `observations`. All of which records consists of a set of original data, together with a set of filters that are to be applied to the original data. A filter could for instance be a describing a subset of the observations together with a scalar that is to be applied to the errors or a subset of the realisations that are considered active etc. In this manner one can describe edits to the underlying data without changing the data itself.

I would then like the user to be able to describe the updating pipeline as follows:
```yml
model_update:
  records:
    es_mda_weights: [4, 2, 1]

  pipeline:
    - auto_correlate_observations(threshold: 0.4)
    - es_mda(es_mda_weights) # or equivalently "es_mda(weights: [4, 2, 1])"
```

Then one can define each of the stages in a stages section (either in a site-configuration or in project or experiment configuration):

```yml
stages:
  auto_correlate_observations:
    input:
      - observations
      - simulated_data
      - threshold -> 0.7 # default to 0.7

    output:
      - observations
      - simulated_data

    records:
      spearman_setup:
        fcluster:
          t: monocrit
          criterion: maxclust
          depth: 2
        linkage:
          method: complete

    pipeline:
      - spearman_correlation(spearman_setup)
      - correlated_observation_scaling(my_super_constant: 0.1)
```

**Comment by markusdregi** [comment]
I've tried to make a general discussion on configuration of the correlation jobs in https://github.com/equinor/semeio/issues/67.

**Comment by xjules** [comment]
> When running the correlated observations scaling after the spearman correlation, the threshold is hard-coded to 0.95. Allow the user to modify this value, e.g. through an argument
> 
> Including many arguments may justify creating an optional config file:
> #46
> #44

Not sure if I understand correctly what threshold do you mention here, ie. is it the number of principal components of observations that fall under T(=0.95) variance?

**Comment by xjules** [comment]
@markusdregi I am not sure that I see that main advantage of such a general stage-based pipeline with open inputs and outputs. Why the user needs to modify something that lies behind the curtain, ie. what data this particular module requires, this should immutable right? For example, we can't exclude observations from `COS` since the job requires observations.

**Comment by markusdregi** [comment]
> For example, we can't exclude observations from COS since the job requires observations. If there is another filter that acts on observations by injecting it before COS will make the observations implicitly filtered on COS input.

From above:

> The records input to the model updating pipeline at large would be the parameters, simulation_results and observations. All of which records consists of a set of original data, together with a set of filters that are to be applied to the original data. A filter could for instance be a describing a subset of the observations together with a scalar that is to be applied to the errors or a subset of the realisations that are considered active etc. In this manner one can describe edits to the underlying data without changing the data itself.

The idea in general would be to use this as a testbed of developing an update pipeline in ERT that would in the future describe the entire updating procedure to be executed.

**Comment by markusdregi** [comment]
This is resolved by #68
